Adding responsible LLM guidelines to Docs (#2616)
terrapower/armi: doc/developer/standards_and_practices.rst

    consistency. Code reviewers should make sure to be familiar with the standards, so that their comments are
    consistent with other reviewers.

Responsible use of Large Language Models (LLMs)
===============================================

> ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation an analysis. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Additionally, ARMI's Contributor License Agreement (CLA) requires that `work submitted is your own <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L23>`_, and you `hold the copyright to any work submitted <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L45>`_. Because an LLM cannot be held responsible or accountable for quality or saftey, cannot be considered original work, and cannot hold copyright, ARMI contributions must adhere to the following requirements:

Reviewer: Here I fix a couple spelling errors.

And I am removing discussion of copyright, because I think it focuses the discussion more to talk about quality and safety (which are crucial in the nuclear space).

Feel free to argue for/against.

Author: I do think that copyright and adhering to the license is important to note, and should be kept in, but I agree the emphasis should be on quality/safety.

> ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation an analysis. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Additionally, ARMI's Contributor License Agreement (CLA) requires that `work submitted is your own <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L23>`_, and you `hold the copyright to any work submitted <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L45>`_. Because an LLM cannot be held responsible or accountable for quality or saftey, cannot be considered original work, and cannot hold copyright, ARMI contributions must adhere to the following requirements:

Reviewer: ```suggestion
ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation and analyses. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Additionally, ARMI's Contributor License Agreement (CLA) requires that `work submitted is your own <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L23>`_, and you `hold the copyright to any work submitted <https://github.com/terrapower/armi/blob/6b5208e8ef95620fb0ce993e000043f3147faabe/CONTRIBUTING.md?plain=1#L45>`_. Because an LLM cannot be held responsible or accountable for quality or saftey, cannot be considered original work, and cannot hold copyright, ARMI contributions must adhere to the following requirements:
```

> 1. All code, comments, and associated documentation generated by an LLM must be reviewed and edited for conciseness and correctness by a human before submitting a GitHub PR or Issue

Reviewer: ```suggestion
1. All code, comments, and associated documentation generated by an LLM must be reviewed and edited for conciseness and correctness by a human before submitting a GitHub PR or Issue.
```

> 2. Tests for code generated by an LLM shall be written by a human.

Reviewer 2: ```suggestion
1. All code, comments, and associated documentation generated by an LLM must be reviewed and edited for conciseness and correctness by a human before submitting a GitHub PR or Issue.
```
Imo. 

Tests fall under the code umbrella. As long as a human reviews and claims ownership of the LLM generated content, I, personally, have no qualms allowing an LLM to write tests.

Author: Having a human write the tests shows they actually understand the code the LLM outputs. Of course an LLM could just write the tests for you, but I'm just not that thrilled to allow a contribution from someone without them deeply on it (as writing a test suite would prove that they did).

This rule is easily avoided but I think the principle of it is important.

Reviewer 2: I disagree, but that's ok. I believe that if a human reviews and claims ownership for tests generated by an LLM, it is fine. 

I've said my peace and will leave you and @john-science to decide how you want to manage. Thanks for engaging!

Reviewer: "Tests for code generated by an LLM shall be written by a human."

This seems like a fine way to ensure that someone has read an understood LLM output.

And it doesn't stop people from using an LLM to _just_ add code coverage tests in a stand-alone PR.

> 3. PR and Issue titles and descriptions shall be written by a human. Fully agentic PRs or Issues will not be accepted.

Reviewer 2: I like this, but I think the second sentence may be hard to be objective about.

Author: @albeanth are you referring to bullet point 3?

Reviewer 2: Yeah sentence two of bullet three.

Author: Sorry for my late response on this @albeanth. While there are cases where this could be difficult to discern, I think emphasizing that we want contributions centered on human understanding is important.

Reviewer 2: Ok

> 3. PR and Issue titles and descriptions shall be written by a human. Fully agentic PRs or Issues will not be accepted.

Reviewer: ```suggestion
3. Fully agentic PRs or Issues will not be accepted. (To that end, PR and Issue titles and descriptions shall be written by a human.)
```

Reviewer: I am just thinking aloud here.

I like the idea of having "Fully agentic PRs or Issues will not be accepted." be at the start here, as it is the important idea.

And I'm just trying to rearrange the text to make that happen.

Thoughts?

    consistency. Code reviewers should make sure to be familiar with the standards, so that their comments are
    consistent with other reviewers.

Responsible use of Large Language Models (LLMs)
===============================================

> ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation an analysis. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Because an LLM cannot be held responsible or accountable for quality or saftey, ARMI contributions must adhere to the following guidelines:

Reviewer: ```suggestion
ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation or analysis. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Because an LLM cannot be held responsible or accountable for quality or saftey, ARMI contributions must adhere to the following guidelines:
```

Reviewer 3: ```suggestion
ARMI is used by TerraPower, LLC and other organizations to perform nuclear reactor simulation or analysis. Because of this, ARMI maintains a strong quality program that requires a human to be held responsible for all changes to the codebase. Because an LLM cannot be held responsible or accountable for quality or safety, ARMI contributions must adhere to the following guidelines:
```

> 1. All code and comments generated by an LLM shall be reviewed by a human before submitting a GitHub PR or Issue.

Reviewer 4: This feels hard to enforce. Maybe the CLA is part of this?

Someone can "review" something generated by the LLM and it still be low-quality / low-effort / wrong, shifting burden to the next person to review the item

Author: @drewj-tp This is true. Can you elaborate on the CLA being "part of this?"

Reviewer 4: All the good stuff has already been said 😅

> 1. All code and comments generated by an LLM shall be reviewed by a human before submitting a GitHub PR or Issue.

Reviewer 5: I'd love to add some language that it should be condensed too. Right now someone could just skim it and post it and not shorten the overly verbose output, which is what we want.

Spitballing: shall be reviewed and edited for conciseness by a human

Author: Make a suggestion comment and I'll add it to the batch!

Reviewer 5: ```suggestion
1. All code and comments generated by an LLM shall be reviewed and edited for conciseness and correctness by a human before submitting a GitHub PR or Issue.
```
Took in drew's comment about correctness

> 2. Documentation for code generated by an LLM shall be written by a human. Documentation generated by an LLM will not be

Reviewer 4: > Documentation generated by an LLM will not be accepted

This feels hard to enforce / detect. In extreme cases it's easy. But I don't think we have the capacity / bandwidth to definitely prove every docs PR does not use an LLM

Reviewer: Prove? No.

But I have seen things that were obviously written by AI already a couple of times:

* A comment started off with an explanation of git and git patches
* Absurdly verbose, multi-thousand-line-long, explanations
* Explanations with just terrible English (though I understand ESL exists, it comes across differently)

And I feel like in all three of those cases I would be happy to be like, "NOPE, robot text, I'm not reading this".

Reviewer 2: I think point 1 and 2 can be combined. Consider:
"_All code, comments, and associated documentation generated by an LLM must be reviewed and edited for conciseness and correctness by a human before submitting a GitHub PR or Issue._" 

I also echo @drewj-tp in that this is difficult to be objective on. The bullets you provided, @john-science, can easily be done by a real human.

I would instead advocate for a "no-AI/LLM/agent bot account contributions" (which I'm unsure I agree with to be honest) and all PRs must follow the contribution guidelines set forth in https://github.com/terrapower/armi/blob/main/doc/developer/tooling.rst#good-pull-requests. 

If we find that a user is using a bot to submit through their account in a malicious/nuisance way, that account can be banned.

Reviewer 3: I think it is possible to spend far too much time debating about the detectability of LLM generated text. I think providing guidelines is an acceptable way to do it, and if a cyborg submits text that is high quality then there is no issue. To use an unfortunate comparison, these rules are like HOA rules: if you don't do something egregious then there will be no problems, but the administrators have a stick to hit you with if you cause them/your neighbors problems.

re LLM accounts: if someone hasn't taken the time to view the commit themselves, I don't see the point of accepting their contributions. E.g. if John wanted to ask $AGENT to improve the code, he can do that himself. My impression is that commits coming from agent accounts are not seen by humans until they are submitted.

Reviewer: > I would instead advocate for a "no-AI/LLM/agent bot account contributions"

@albeanth The idea @yardasol had was to say "no fully agentic" contributions to ARMI will be allowed. What do you think of that?

Reviewer 2: How would one define a "fully agentic contribution"? If I use an agent to create a patch, review said patch (and like it as is with no edits), apply it to a feature branch, and create a PR, is that a full agentic contribution? Or is this more for bot accounts with no human-in-the-loop? For the latter, are those accounts easily/objectively identified?

Author: @albeanth I think the workflow you described is reasonable, since you are reviewing the code and adhering to point 1. The main concern we had with that bullet point was low-effort/fully automated PRs without any human eyes being on it.

Author: > > Documentation generated by an LLM will not be accepted
> 
> This feels hard to enforce / detect. In extreme cases it's easy. But I don't think we have the capacity / bandwidth to definitely prove every docs PR does not use an LLM

We definitely don't have the bandwidth to prove every docs PR is human generated rather than LLM generated, especially as models are refined and become more sophisticated. A question for further consideration: is it worth having this guideline/rule if it's not easily enforcable / violations aren't easily detectable?

Reviewer 3: >A question for further consideration: is it worth having this guideline/rule if it's not easily enforcable / violations aren't easily detectable?

It is absolutely worth having a guideline that lets the maintainers feel empowered to summarily reject PRs (true with or without LLMs). Remember, this is an open source project and the goal _should not_ be to have rules that are as leak proof as a law or a contract. The threshold that should be aimed for is far lower, and the goal is to protect the time of the maintainers. Trying to make a technical/contract-like solution for a social issue is a fool's errand and I implore everyone in this thread to keep that in mind.

   accepted.
3. Tests for code generated by an LLM shall be written by a human, with an exception made for PRs expanding test

>    coverage submitted by a qualified developer. For all other cases tests generated by an LLM will not be accepted.

Reviewer 3: Suggestion: clarify what a "qualified developer" is, perhaps specifically stating that nobody outside the TP QA program (credit to Drew on that phrasing) is allowed to submit LLM generated tests.

Reviewer: I think we could just remove the phrase "by a qualified developer" here. And let the sentence stand.

Author: I agree perhaps it should be removed

Reviewer 2: I don't understand this line. What constitutes test code "expanding test coverage"? What's to say said contribution isn't garbage/reviewed by a human prior to submittal? Tests are equally as important to non-test changes, in my opinion. I think this contradicts point 1 and should be considered for removal altogether with.

Author: In my mind, having a human write the tests proves that they actually understand code the LLM wrote. Obviously, if a PRs intention is to just expand test coverage, then the LLM would just be writing tests so there's not much work for the human to do other than review the tests and understand them.

I agree the way around this would be to force ALL tests to be written by a human. The question then is, is that reasonable?

Reviewer: @albeanth At some point, Olek rewrote this bullet point to:

```
2. Tests for code generated by an LLM shall be written by a human.
```

So that, I think, is much more clear.

4. PR and Issue titles and descriptions shall be written by a human. Agentic PRs or Issues will not be accepted.

> 

Reviewer 4: What happens if we find / suspect violations?

1. Will we close an issue / PR outright?
2. Will we block frequent violators?

Reviewer: Yeah, this seems important to point out.

> 1. Will we close an issue / PR outright?

I am in favor of just closing Sloppy comments and PRs. If everyone is okay with that, I am happy to add that to the docs.

> 2. Will we block frequent violators?

Oh yeah we will. But I don't know if we need to call this out in the docs. We have had to ban several user accounts from ARMI for various unpleasant reasons. I don't know if it's very fun to try to describe them all in docs. Just unpleasant stuff.

Just my opinions. Thoughts?

Reviewer: My notion is after the bullet point list we just put in a simple sentence saying something like:

"PRs and Issues that violate the above rules may be summarily closed."

Reviewer 2: > "PRs and Issues that violate the above rules may be summarily closed."

I like this. I think a solution is to beef up the rules for PRs and Issues. Go from "Good Pull Requests" guidance / etiquette to "This is the required list to create a PR or Issue. If you don't meet these things, it will be closed." This would provide a more objective checklist for both humans and bots/agents add hopefully raise the bar for _all_ contributions.

Author: > * Will we close an issue / PR outright?

Yes
> * Will we block frequent violators?

Yes.

@john-science I like this suggestion.
@albeanth So you're saying we add this to the dev docs and also add a checkpoint to the PR template?

Reviewer 2: Yes to the dev docs. Unsure about a checkpoint in the PR template. Maybe a reminder that's in the comments, but idk how useful that is.

Reviewer: Thanks for the change!
